### DISCUSSION TITLE
How come the neopixel lib works on the feather tft with esp32-s3 however it doesn't work on any of the m5 boards like atomS3  and coreS3  which also have a esp32-s3

### DISCUSSION THREAD

**Discussion by kairos0ne** [Moddable-OpenSource/moddable#1187] [General]
I asked chatgpt a similar question and its response was: 

>The Neopixel library is specifically designed to control Neopixel or WS2812 LED strips, and it requires hardware support for these LEDs. If your board does not have built-in Neopixel support, you won't be able to use the Neopixel library directly.

Which I feel must be nonsense right, surely these boards are using the same chip. Why would one not write the lib for all boards as these neopixel strips are external hardware devices that you connect via pins. It makes no sense to only support boards that have an on board neopixel.

**Comment by phoddie** [comment]
I've successfully used Neopixels on several ESP32-S3 devices. There may be a conflict or misconfiguration getting in the way here. To be able to help, some additional information is going to be needed:

- What code are you using to configure the Neopixel?
- Which GPIO is the Neopixel strand connected to?
- How is it failing? Answers might include: throwing an exception, not changing the LEDs, and hanging the ESP32-S3.

**Comment by kairos0ne** [comment]
Sure np, 

Here is main.js I adapted from uart-server example. 

```javascript
import { BLEServer } from "bleserver";
import { uuid, typedValueToBuffer } from "btutils";
import Timer from "timer";
import NeoPixel from 'neopixel';


// UUIDs for the UART service and its characteristics
const SERVICE_UUID = uuid`6E400001-B5A3-F393-E0A9-E50E24DCCA9E`;
const CHARACTERISTIC_RX_UUID = uuid`6E400002-B5A3-F393-E0A9-E50E24DCCA9E`;
const CHARACTERISTIC_TX_UUID = uuid`6E400003-B5A3-F393-E0A9-E50E24DCCA9E`;

const timeout = 30000; // 30 seconds
const duration = 0; // unlimited



const UART_SERVER = "UART Server";

// pin for the neopixel
const neopixelPin = 6

let neopixel = new NeoPixel({ length: 12, pin: neopixelPin, order: "GRB" });


const red = neopixel.makeRGB(255, 0, 0);
const yellow = neopixel.makeRGB(255, 255, 0);
const orange = neopixel.makeRGB(255, 165, 0);
const purple = neopixel.makeRGB(128, 0, 128);
const pink = neopixel.makeRGB(255, 192, 203);
const cyan = neopixel.makeRGB(0, 255, 255);
const magenta = neopixel.makeRGB(255, 0, 255);
const green = neopixel.makeRGB(0, 255, 0);
const blue = neopixel.makeRGB(0, 0, 255);
const white = neopixel.makeRGB(255, 255, 255);
const black = neopixel.makeRGB(0, 0, 0);

function setAll(color) {
	for (let i = 0; i < neopixel.length; i++) {
		neopixel.setPixel(i, color);
		neopixel.update();
	}
}

function clearAll() {
	setAll(black);
}

function setAllColor(color) {
	setAll(color);
}

class UARTServer extends BLEServer {

	onReady() {
		this.deviceName = "VEGAONE";
		this.onDisconnected();
	}
	onConnected() {
		this.stopAdvertising();
	}
	onCharacteristicNotificationEnabled(characteristic) {
		if (characteristic.uuid.equals(CHARACTERISTIC_TX_UUID)) {
			this.tx = characteristic;
			this.timer = Timer.repeat(id => {
				if (this.tx_characteristic) {
					this.tx.notify("Hello from UART Server\n");
				}
			}, 1000);
		} else if (characteristic.uuid.equals(CHARACTERISTIC_RX_UUID)) {
			this.rx = characteristic;
		}

	}

	onCharacteristicNotificationDisabled(characteristic) {
		if (characteristic.uuid.equals(CHARACTERISTIC_TX_UUID)) {
			Timer.clear(this.timer);
			this.timer = null;
			this.disconnect();
		}
	}
	onDisconnected() {
		this.tx = null;
		this.rx = null;
		// remove the timer
		if (this.timer) {
			Timer.clear(this.timer);
			this.timer = null;
		}
		this.startAdvertising({
			advertisingData: { flags: 6, completeName: this.deviceName, completeUUID128List: [SERVICE_UUID] }
		});
	}

	onCharacteristicWritten(characteristic, value) {

		if (characteristic.uuid.equals(CHARACTERISTIC_RX_UUID)) {
			trace.left(value, UART_SERVER);

			if (value == 1) {
				setAllColor(black);
			}

			if (value == 2) {
				setAllColor(red);
			}

			if (value == 3) {
				setAllColor(yellow);
			}

			if (value == 4) {
				setAllColor(orange);
			}

			if (value == 5) {
				setAllColor(purple);
			}

			if (value == 6) {
				setAllColor(pink);
			}

			if (value == 7) {
				setAllColor(cyan);
			}

			if (value == 8) {
				setAllColor(magenta);
			}

			if (value == 9) {
				setAllColor(green);
			}

			if (value == 10) {
				setAllColor(blue);
			}

			if (value == 11) {
				setAllColor(white);
			}

			if (value == 12) {
				setAllColor(black);
			}

			if (value == 13) {
				clearAll();
			}

		}
	}

}
let server = new UARTServer;

```

Im using 6 which which g6 on the back of  the atom image attached. 

<img width="604" alt="image" src="https://github.com/Moddable-OpenSource/moddable/assets/15980261/5b861ae6-5dfb-48ac-8b5b-54e0d9181bd3">

and I get: 

```
# Break: (host): module "neopixel" not found!
```

From XSBUG

<img width="1609" alt="image" src="https://github.com/Moddable-OpenSource/moddable/assets/15980261/20828c1f-3aae-4379-ac8b-b4ab5440847e">

**Comment by kairos0ne** [comment]
I also tried G1 with a grove connector in Port A of the atom S3, same result - worth noting that I get the same for the core s3. But works really well on the feather tft esp32-s3 from adafruit. However as you noted Im probably missing some configuration. I note that at least by the code in the module I believe its supposed to be supported on esp32 using a * wildcard path. I'm very new to this and my my experience is limited. Hope that helps.

**Comment by mkellner** [comment]
Hello. I tried the examples/drivers/neopixel app with esp32/m5atom_s3 pin 6.


https://github.com/Moddable-OpenSource/moddable/assets/6822704/76439d45-00f4-425e-9ebc-2d6092840cad

**Comment by kairos0ne** [comment]
Ja, it works very well when you import the lib from manifest.json. 0__o such a silly thing to overlook. Thanks @mkellner
